common/src/main/java/com/wynntils/features/user/ChatTabsFeature.java

```diff
@@ -0,0 +1,100 @@
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.TypeOverride;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.core.managers.Model;
+import com.wynntils.gui.widgets.ChatTabButton;
+import com.wynntils.mc.event.ScreenInitEvent;
+import com.wynntils.mc.event.ScreenOpenedEvent;
+import com.wynntils.mc.event.ScreenRenderEvent;
+import com.wynntils.wynn.event.ChatMessageReceivedEvent;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+import net.minecraft.client.gui.screens.ChatScreen;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatTabsFeature extends UserFeature {
+    public static ChatTabsFeature INSTANCE;
+
+    @Config(visible = false)
+    public List<ChatTab> chatTabs = Arrays.asList(
+            new ChatTab("All", true, Sets.newHashSet(RecipientType.values()), null),
+            new ChatTab("Global", true, Sets.newHashSet(RecipientType.GLOBAL), null),
+            new ChatTab("Local", true, Sets.newHashSet(RecipientType.LOCAL), null),
+            new ChatTab("Guild", true, Sets.newHashSet(RecipientType.GUILD), null),
+            new ChatTab("Party", true, Sets.newHashSet(RecipientType.PARTY), null),
+            new ChatTab("Private", true, Sets.newHashSet(RecipientType.PRIVATE), null),
+            new ChatTab("Shout", true, Sets.newHashSet(RecipientType.SHOUT), null));
+
+    @TypeOverride
+    public Type chatTabsType = new TypeToken<List<ChatTab>>() {}.getType();
+
+    @Override
+    public List<Class<? extends Model>> getModelDependencies() {
+        return List.of(ChatModel.class, ChatTabModel.class);
+    }
+
+    // We do this here, and not in ChatTabModel to not introduce a feature-model dependency.
+    @SubscribeEvent(priority = EventPriority.LOWEST)
+    public void onChatReceived(ChatMessageReceivedEvent event) {
+        // Firstly, find the FIRST matching tab with high priority
+        for (ChatTab chatTab : chatTabs) {
+            if (chatTab.isLowPriority()) continue;
+
+            if (chatTab.matchMessageFromEvent(event)) {
+                ChatTabModel.addMessageToTab(chatTab, event.getMessage());
+                return;
+            }
+        }
+
+        // Secondly, match ALL tabs with low priority
+        for (ChatTab chatTab : chatTabs) {
+            if (!chatTab.isLowPriority()) continue;
+
+            if (chatTab.matchMessageFromEvent(event)) {
+                ChatTabModel.addMessageToTab(chatTab, event.getMessage());
+            }
+        }
+    }
+
+    @SubscribeEvent
+    public void onScreenInit(ScreenInitEvent event) {
+        if (event.getScreen() instanceof ChatScreen chatScreen) {
+            int xOffset = 0;
+
+            for (ChatTab chatTab : chatTabs) {
+                chatScreen.addRenderableWidget(new ChatTabButton(xOffset + 2, chatScreen.height - 35, 40, 13, chatTab));
+                xOffset += 43;
```

Comment:
Should the tab offset depend on the length of the name of the tab?

Reply:
Long names just overflow the button, as it is in legacy. It is rather messy to impl a gui button with a dynamic size. So for now, this should work.

Reply:
Ok. Legacy compatiblity is a perfectly fine excuse for just about anything! :-) (I'm kind of seeing the xkcd "Compiling!" "Oh, okay, carry on then" strip in front of me...)